File: apps/client/app/features/admin/types/SecurityAlert.ts
export type SecurityAlertRuleId =
  | 'repeated_permission_denied'
  | 'multi_resource_permission_probe'
  | 'repeated_policy_block';

export type SecurityAlertSeverity = 'medium' | 'high';
export type SecurityAlertStatus = 'open' | 'acknowledged' | 'resolved';
export type SecurityAlertResolutionType =
  | 'mitigated'
  | 'false_positive'
  | 'accepted_risk';
export type SecurityAlertActorState =
  | 'active'
  | 'suspended'
  | 'removed'
  | 'deleted';

export const SECURITY_ALERT_OPERATIONS = {
  LIST: 'security_alert.list',
  SUMMARY: 'security_alert.summary',
  DETAIL: 'security_alert.detail',
  EVIDENCE_LIST: 'security_alert.evidence.list',
  ACKNOWLEDGE: 'security_alert.acknowledge',
  RESOLVE: 'security_alert.resolve',
  REOPEN: 'security_alert.reopen',
} as const;

export type SecurityAlertOperation =
  (typeof SECURITY_ALERT_OPERATIONS)[keyof typeof SECURITY_ALERT_OPERATIONS];

export function isSecurityAlertOperation(
  value: string,
): value is SecurityAlertOperation {
  return (Object.values(SECURITY_ALERT_OPERATIONS) as string[]).includes(value);
}

export type SecurityAlertSafeActor = {
  id: string | null;
  display_name: string | null;
  state: SecurityAlertActorState;
};

export type SecurityAlertListItem = {
  id: string;
  organization_id: string;
  rule_id: SecurityAlertRuleId;
  rule_version: string;
  severity: SecurityAlertSeverity;
  status: SecurityAlertStatus;
  policy_reason: string | null;
  actor: SecurityAlertSafeActor;
  occurrence_count: number;
  first_detected_at: string;
  last_detected_at: string;
  version: number;
  created_at: string;
  updated_at: string;
};

export type SecurityAlertAcknowledgement = {
  by: SecurityAlertSafeActor;
  at: string;
};

export type SecurityAlertResolution = {
  type: SecurityAlertResolutionType;
  reason: string;
  by: SecurityAlertSafeActor;
  at: string;
};

export type SecurityAlertDetail = SecurityAlertListItem & {
  evidence_count: number;
  acknowledged: SecurityAlertAcknowledgement | null;
  resolution: SecurityAlertResolution | null;
};

export type SecurityAlertListResponse = {
  total: number;
  items: SecurityAlertListItem[];
};

export type SecurityAlertSummaryItem = {
  id: string;
  rule_id: SecurityAlertRuleId;
  severity: SecurityAlertSeverity;
  actor: SecurityAlertSafeActor;
  occurrence_count: number;
  last_detected_at: string;
};

export type SecurityAlertSummaryResponse = {
  open_count: number;
  high_open_count: number;
  recent_items: SecurityAlertSummaryItem[];
};

export type SecurityAlertAuditLogItem = {
  id: string;
  occurred_at: string;
  actor_id: string | null;
  actor_type: string;
  category: string;
  action: string;
  target_type: string | null;
  target_id: string | null;
  status: 'success' | 'failure';
  request_id: string | null;
  required_permission?: string | null;
  requested_operation?: SecurityAlertOperation | null;
  denial_reason?: string | null;
};

export type SecurityAlertAuditLogListResponse = {
  total: number;
  items: SecurityAlertAuditLogItem[];
};

export type SecurityAlertListParams = {
  page?: number;
  limit?: number;
  severity?: SecurityAlertSeverity;
  status?: SecurityAlertStatus;
  ruleId?: SecurityAlertRuleId;
  actorId?: string;
  startAt?: string;
  endAt?: string;
};

export type ResolveSecurityAlertInput = {
  expectedVersion: number;
  resolutionType: SecurityAlertResolutionType;
  reason: string;
};

File: apps/shared/schemas/security_alert.py
from datetime import datetime
from typing import Literal
from uuid import UUID

from apps.shared.db.models.audit_log import ActorType, AuditCategory, AuditStatus
from apps.shared.schemas.member_access import normalize_management_reason
from pydantic import BaseModel, ConfigDict, Field, field_validator

SecurityAlertRuleId = Literal[
    "repeated_permission_denied",
    "multi_resource_permission_probe",
    "repeated_policy_block",
]
SecurityAlertSeverity = Literal["medium", "high"]
SecurityAlertStatus = Literal["open", "acknowledged", "resolved"]
SecurityAlertResolutionType = Literal[
    "mitigated", "false_positive", "accepted_risk"
]
SecurityAlertActorState = Literal["active", "suspended", "removed", "deleted"]


class SecurityAlertSafeActor(BaseModel):
    # Lifecycle actor FK는 user 삭제 시 SET NULL이므로 historical projection은
    # ID 없이 deleted 상태만 남을 수 있다. Alert subject actor는 항상 ID가 있다.
    id: UUID | None
    display_name: str | None
    state: SecurityAlertActorState


class SecurityAlertListItem(BaseModel):
    id: UUID
    organization_id: UUID
    rule_id: SecurityAlertRuleId
    rule_version: str
    severity: SecurityAlertSeverity
    status: SecurityAlertStatus
    policy_reason: str | None = None
    actor: SecurityAlertSafeActor
    occurrence_count: int = Field(ge=0)
    first_detected_at: datetime
    last_detected_at: datetime
    version: int = Field(ge=1)
    created_at: datetime
    updated_at: datetime


class SecurityAlertAcknowledgement(BaseModel):
    by: SecurityAlertSafeActor
    at: datetime


class SecurityAlertResolution(BaseModel):
    type: SecurityAlertResolutionType
    reason: str
    by: SecurityAlertSafeActor
    at: datetime


class SecurityAlertDetail(SecurityAlertListItem):
    evidence_count: int = Field(ge=0)
    acknowledged: SecurityAlertAcknowledgement | None = None
    resolution: SecurityAlertResolution | None = None


class SecurityAlertListResponse(BaseModel):
    total: int = Field(ge=0)
    items: list[SecurityAlertListItem]


class SecurityAlertSummaryItem(BaseModel):
    id: UUID
    rule_id: SecurityAlertRuleId
    severity: SecurityAlertSeverity
    actor: SecurityAlertSafeActor
    occurrence_count: int = Field(ge=0)
    last_detected_at: datetime


class SecurityAlertSummaryResponse(BaseModel):
    open_count: int = Field(ge=0)
    high_open_count: int = Field(ge=0)
    recent_items: list[SecurityAlertSummaryItem] = Field(max_length=5)


class SecurityAlertAuditLogItem(BaseModel):
    id: UUID
    occurred_at: datetime
    actor_id: UUID | None
    actor_type: ActorType
    category: AuditCategory
    action: str
    target_type: str | None
    target_id: str | None
    status: AuditStatus
    request_id: str | None = None
    required_permission: str | None = None
    requested_operation: str | None = None
    denial_reason: str | None = None


class SecurityAlertAuditLogListResponse(BaseModel):
    total: int = Field(ge=0)
    items: list[SecurityAlertAuditLogItem]


class SecurityAlertVersionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")

    expected_version: int = Field(ge=1)


class SecurityAlertAcknowledgeRequest(SecurityAlertVersionRequest):
    pass


class SecurityAlertReopenRequest(SecurityAlertVersionRequest):
    pass


class SecurityAlertResolveRequest(SecurityAlertVersionRequest):
    resolution_type: SecurityAlertResolutionType
    reason: str

    @field_validator("reason")
    @classmethod
    def normalize_reason(cls, value: str) -> str:
        normalized = normalize_management_reason(value)
        if normalized is None:
            raise ValueError("reason must not be blank")
        return normalized
